src/System.Management.Automation/engine/runtime/Binding/Binders.cs

```diff
@@ -2029,36 +2029,38 @@ private static object NullRule(CallSite site, object obj)
 
         internal static bool IsValueTypeMutable(Type type)
         {
-            if (type.IsPrimitive || type.IsEnum)
+            // First, check for enums/primitives and compiler-defined attributes.
+            if (type.IsPrimitive
+                || type.IsEnum
+                || type.IsDefined(typeof(System.Runtime.CompilerServices.IsReadOnlyAttribute), inherit: false))
             {
                 return false;
             }
 
-            // If there are any fields, the type is mutable.
-            if (type.GetFields(BindingFlags.Public | BindingFlags.Instance).Any())
+            // If the builtin attribute is not present, check for a custom attribute from by the compiler. If the
+            // library targets netstandard2.0, the compiler can't be sure the attribute will be provided by the runtime,
+            // and defines its own attribute of the same name during compilation.
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
             {
-                return true;
+                if (attribute.GetType().FullName.Equals(
+                    "System.Runtime.CompilerServices.IsReadOnlyAttribute",
+                    StringComparison.Ordinal))
```

Comment:
What's the reason for needing to do a type name comparison rather than a simple type comparison here?

Reply:
> What's the reason for needing to do a type name comparison rather than a simple type comparison here?

@rjmholt if you target `netstandard2.0` and use `readonly struct`s then Roslyn embeds the attribute definition in the assembly, so type identity will only be reliable for assemblies targeting netcoreapp3.0+ (approximately, don't remember when that was introduced).

Reply:
Ok got it -- might be worth a comment so that someone doesn't "optimise" it

Reply:
I already added comments for this just above (2040-2043) - should I move those closer to this line?

Reply:
Ah, I would just add something like "so we must check for the type by name rather than by reference" after the bit about the compiler